### ISSUE TITLE
Format of inputs for MVDR beamformer

### ISSUE THREAD

**Issue by danaiksez** [speechbrain/speechbrain#869]
Hi,

I'm trying to apply the MVDR beamformer to a two-channel input recording. I follow the example use case in _multi_mic.py_, however it is unclear to me what the _xs_noise_ variable should contain. Should it be the noise diffuse from the audio recording that I have to separate somehow as a first step in advance? In this case, is the _xs_speech_ variable expected to contain both channels of the input recording, or is it something else that I am missing there?

Thanks a lot!

**Comment by mravanelli** [comment]
Thank you for the question. This is something for @FrancoisGrondin.

**Comment by FrancoisGrondin** [comment]
Hi,

Here is a code sample that shows how to use the MVDR module. Please let me know if this helps understanding how to use the module.

```from speechbrain.processing.features import STFT, ISTFT
from speechbrain.processing.multi_mic import Covariance
from speechbrain.processing.multi_mic import GccPhat
from speechbrain.processing.multi_mic import Mvdr

import matplotlib.pyplot as plt
import torch

xs_speech = read_audio('speech_-0.82918_0.55279_-0.082918.flac') # [time, channels]
xs_speech = xs_speech.unsqueeze(0) # [batch, time, channels]
xs_noise_diff = read_audio('noise_diffuse.flac') # [time, channels]
xs_noise_diff = xs_noise_diff.unsqueeze(0) # [batch, time, channels]
xs_noise_loc = read_audio('noise_0.70225_-0.70225_0.11704.flac') # [time, channels]
xs_noise_loc =  xs_noise_loc.unsqueeze(0) # [batch, time, channels]
fs = 16000 # sampling rate

ss = xs_speech
nn_diff = 0.05 * xs_noise_diff
nn_loc = 0.05 * xs_noise_loc
xs_diffused_noise = ss + nn_diff
xs_localized_noise = ss + nn_loc

stft = STFT(sample_rate=fs)
cov = Covariance()
gccphat = GccPhat()
mvdr = Mvdr()
istft = ISTFT(sample_rate=fs)

Xs = stft(xs_diffused_noise)
XXs = cov(Xs)
tdoas = gccphat(XXs)
Ys_mvdr = mvdr(Xs, XXs, tdoas)
ys_mvdr = istft(Ys_mvdr)

plt.figure(1)
plt.title('Noisy signal at microphone 1')
plt.imshow(torch.transpose(torch.log(Xs[0,:,:,0,0]**2 + Xs[0,:,:,1,0]**2), 1, 0), origin="lower")
plt.figure(2)
plt.title('Noisy signal at microphone 1')
plt.plot(xs_diffused_noise.squeeze()[:,0])
plt.figure(3)
plt.title('Beamformed signal')
plt.imshow(torch.transpose(torch.log(Ys_mvdr[0,:,:,0,0]**2 + Ys_mvdr[0,:,:,1,0]**2), 1, 0), origin="lower")
plt.figure(4)
plt.title('Beamformed signal')
plt.plot(ys_mvdr.squeeze())
plt.show()```

**Comment by danaiksez** [comment]
Thank you for the quick response!
This is indeed the use case example that I was trying to run, giving as input a two-channel audio recording. Therefore, I'm not sure what input should be given to the _xs_noise_diff_ and _xs_noise_loc_ variables, since I only have a single audio recording and not any kind of a noise file. Any clarification on that would be really helpful.

**Comment by FrancoisGrondin** [comment]
Oh I see! So usually in this case you want to have a method to roughly separate speech from noise (it doesn't have to be perfect, but you want to capture some of it). So what you can do is use a time-frequency mask estimation on your mixture (there are many tools with speechbrain to achieve this), and then you generate one signal with mostly noise, and the other one with mostly speech. You then feed this to MVDR. You might wonder why not just take the signals obtained from masking the mixture? The big plus with beamforming based separation is that you do not have non-linear distortion.

Does this help?

**Comment by danaiksez** [comment]
Thanks, this cleared things up a lot!

In order to separate speech from noise, would Speechbrain's pretrained SepFormer (https://huggingface.co/speechbrain/sepformer-wsj02mix) help towards this direction, or is there any other way that you would recommend?

**Comment by FrancoisGrondin** [comment]
Yes, SepFormer is a good choice, as long as all sources are speech signals. If you want to do noise/speech separation, you are looking for speech enhancement methods, and I think this tutorial could get you started: https://colab.research.google.com/drive/18RyiuKupAhwWX7fh3LCatwQGU5eIS3TR?usp=sharing#scrollTo=pYKnhdQK8oGy

Cheers :)

**Comment by danaiksez** [comment]
Thanks a lot!
